---
title: "AD_Synapse_snRNAseqPFC"
author: "Lee Marshall"
date: "2021_08_11 - `r format(Sys.time(), '%Y_%m_%d')`"
output: 
  html_document:
    code_folding: show
    self_contained: yes
    toc_float: TRUE
    theme: yeti
---


# AD_Synapse_snRNAseqPFC {.tabset .tabset-fade .tabset-pills}

PAPER  
"Single-cell transcriptomic analysis of Alzheimer’s disease"  
15 early AD, 9 late AD, 24 healthy control PFC brain nuclei
Downloaded from SYNAPSE  


## Packages_Session
  
```{r Packages_Session}
#---------- Knitr
setwd("/Volumes/projects_secondary/bras/Lee/ASAP/AD_Synapse_snRNAseqPFC")
knitr::opts_chunk$set(warning=FALSE, message=FALSE, cache=FALSE, error=FALSE, cache.lazy=FALSE)
#knitr::opts_knit$set(root.dir='/Volumes/projects_secondary/bras/Lee/ASAP/Aging_Mice_GSE129788')
#options(scipen=999)
#options(stringsAsFactors = FALSE)
# Chunk Options  
# include = FALSE, runs code, prevents code and results from appearing  
# echo = FALSE, runs code, prevents code from appearing but not the results  
# eval = FALSE, does not run code  
# message = FALSE prevents messages that are generated by code from appearing 
# warning = FALSE prevents warnings that are generated by code from appearing 
# fig.cap = "..." adds a caption to graphical results.

#---------- Packages
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("fgsea")
#install.packages('metap')
suppressPackageStartupMessages({
  library(betareg)
  library(biomaRt)
  library(corrplot)
  library(cowplot)
  library(DoubletFinder)
  library(dplyr)
  library(DT)
  library(fgsea)
  library(ggplot2)
  library(ggpubr)
  library(gprofiler2)
  library(harmony)
  library(knitr)
  library(kableExtra)
  library(lubridate)
  library(magrittr)
  library(MAST)
  library(metap)
  library(multtest)
  library(parallel)
  library(purrr)
  library(RColorBrewer)
  library(readr)
  library(reshape2)
  library(reticulate)
  library(rmarkdown)
  library(Seurat)
  library(stringr)
  library(tibble)
  library(tidyr)
  library(viridis)
})
sessionInfo()
#  library(reticulate)
#  use_condaenv("base")
#  use_python("/Users/lee.marshall/opt/anaconda3/bin/python", required = TRUE)
#  display python config
#  print(py_config())
#  display module search path
#  sys <- import("sys")
#  sys$path
#  py_module_available("scanpy")
```


## Cellranger
  
cellranger-6.0.1 to align and count  
performed mRNA alignment for standard single cell counts  
also performed mRNA and premRNA (introns) counts for trajectory analysis  

```{bash fastq, eval=FALSE}
#---------- Download fastq files
#----- SRAtoolkit
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140231
qsubs.sh -a srr_names -n sratoolkit_prefetch -s "prefetch -X 20G SAMPLE"
qsubs.sh -a srr_names -n sartoolkit_fastqdump -s "fastq-dump --split-files /home/lee.marshall/bras_primary/software/sratoolkit/ncbi/sra/SAMPLE.sra --outdir fastq"
#----- Rename fastq samples
while read line ; do mv $(awk '{print $1,$2}' <<<$line) ; done < names.txt
#----- gzip all fastq files
qsubs.sh -a names.txt -n gzip_fastq -s "gzip SAMPLE"
```

```{bash Cellranger, eval=FALSE}
#---------- cellranger_mRNA
module load cellranger/cellranger-6.0.1
qsubs.sh -n cellranger_count -c 4 -w 48 -a names.txt -s "cellranger count --sample=SAMPLE --id=SAMPLE_count --localcores=4 --fastqs=./fastq --chemistry=SC3Pv2 --transcriptome=/home/lee.marshall/bras_primary/software/cellranger_v6/refdata-gex-GRCh38-2020-A"

#---------- cellranger_mRNA_premRNA
module load cellranger/cellranger-6.0.1
qsubs.sh -n cellranger_count -c 4 -w 48 -a ../names.txt -q labrie -s "cellranger count --sample=SAMPLE --id=SAMPLE_count --localcores=4 --fastqs=../fastq --chemistry=SC3Pv2 --include-introns  --transcriptome=/home/lee.marshall/bras_primary/software/cellranger_v6/refdata-gex-GRCh38-2020-A"

```

```{r Cellranger_Summary, echo=FALSE, fig.cap="Cellranger_Summary", out.width='50%', eval=FALSE}
knitr::include_graphics("counts/OX1X_count/OX1X_count_summary.png")
knitr::include_graphics("counts/OX1X_count/OX1X_count_analysis.png")
```



## Seurat QC 
  
Quality control graphs used to identify filtering thresholds  
Senescence Cell not selectively removed  
  
```{r Seurat_QC, eval=FALSE}
#---------- Seurat_Object
#---- Counts cellranger_mRNA_premRNA
covar <- read_csv("paper/snRNAseqPFC_BA10_covar.csv", 
                  col_types = cols(PROJECT_ID = col_character()))
dir <- "cellranger_mRNA_premRNA/"
name <- covar$SAMPLE

#---- Count Matrix
seurat_counts <- lapply(name, function(i){
  count <- Read10X(data.dir=paste0(dir,i,"_count/outs/filtered_feature_bc_matrix/"),
                  gene.column=2, # col1 Ensemble, col2 Symbol
                  unique.features=TRUE,
                  strip.suffix=TRUE) 
  seurat_obj <- CreateSeuratObject(counts=count, 
                                   project=i,
                                   min.cells=10,
                                   min.genes=200)
})

#---- Merge Seurat Objects
seurat_combined <- merge(x = seurat_counts[[1]], 
                               y = seurat_counts[2:length(seurat_counts)],
                               add.cell.ids = unlist(name), 
                               project="AD_Synapse_snRNAseqPFC")

#---------- Add Metadata
# View(seurat_combined@meta.data)
# orig.ident: contains the sample identity if known
# nCount_RNA: number of UMIs per cell
# nFeature_RNA: number of genes detected per cell
#---- Create metadata dataframe
metadata <- seurat_combined@meta.data
#---- Add cell IDs to metadata
metadata$cells <- rownames(metadata)
#---- Add number of genes per UMI for each cell to metadata
metadata$log10GenesPerUMI <- log10(metadata$nFeature_RNA) / log10(metadata$nCount_RNA)
#---- Compute percent mito ratio
metadata <- cbind(metadata,
                  PercentageFeatureSet(object = seurat_combined, pattern = "^MT-") / 100)
names(metadata)[6] <- "mitoRatio"
#---- Rename columns
names(metadata)[1:3] <- c("sample","nUMI","nGene")
#---- Create Sample
covar$sample <- covar$SAMPLE
metadata <- metadata %>% 
              left_join(covar, by = "sample") %>% 
              dplyr::rename(group = AD_GROUP)
#---- Add metadata back to Seurat object
rownames(metadata) <- metadata$cells
seurat_combined@meta.data <- metadata
#---- Senescence Markers, p16 = CDKN2A, p21 = CDKN1A
seurat_combined$CDKN2A_Count <- 
  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == "CDKN2A",]
seurat_combined$CDKN1A_Count <- 
  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == "CDKN1A",]
#---- rm data
rm(seurat_counts, metadata)
#---- save rds
saveRDS(seurat_combined, file="seurat_results/seurat_combined.rds")
```

```{r Seurat_QC_Figures}
#---------- Cell_Counts
#----- Load data
seurat_combined <- readRDS(file = "seurat_results/seurat_combined.rds")
#----- Visualize the number of cell counts per sample
seurat_combined@meta.data %>% 
  ggplot(aes(x=sample, fill=sample)) + 
  geom_bar() + 
  ggtitle("Number of Cells") + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Visualize the number of cell counts per group
seurat_combined@meta.data %>% 
  dplyr::count(group, sample) %>% 
  dplyr::rename(Count = n) %>% 
  ggplot(aes(x=group, y=Count, fill=group)) + 
  geom_boxplot() + 
  ggtitle("Number of Cells") + 
  scale_fill_brewer(palette="Dark2") +
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))

#---------- UMI_Counts_per_cell (transcripts)
#----- Visualize the number UMIs/transcripts per cell
seurat_combined@meta.data %>% 
  ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of UMIs/Transcripts per Cell") + 
  ylab("Cell density") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Visualize the number UMIs/transcripts per cell
seurat_combined@meta.data %>% 
  ggplot(aes(color=group, x=nUMI, fill= group)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of UMIs/Transcripts per Cell") + 
  ylab("Cell density") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))

#---------- Genes_per_cell
#----- Visualize the distribution of genes detected per cell via histogram
seurat_combined@meta.data %>% 
  ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 250, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 5000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of Genes per Cell") + 
  scale_x_log10() + 
  theme_classic() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Visualize the distribution of genes detected per cell via histogram
seurat_combined@meta.data %>% 
  ggplot(aes(color=group, x=nGene, fill= group)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 250, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 5000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of Genes per Cell") + 
  scale_x_log10() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Visualize the distribution of genes detected per cell via boxplot
seurat_combined@meta.data %>% 
  ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  geom_boxplot() + 
  ggtitle("Number of Genes vs Number of Cells") +   
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Visualize the distribution of genes detected per cell via boxplot
seurat_combined@meta.data %>% 
  ggplot(aes(x=group, y=log10(nGene), fill=group)) + 
  geom_boxplot() + 
  ggtitle("Number of Genes vs Number of Cells") +   
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(legend.position = "none") +
  theme(plot.title = element_text(hjust=0.5, face="bold"))
#----- Plots
#cowplot::plot_grid(p1, p2, ncol=2)

#---------- UMI_Counts_vs_Genes
#----- Visualize the correlation between genes detected and number of UMIs/transcripts 
# determine whether strong presence of cells with low numbers of genes/UMIs
seurat_combined@meta.data %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point(size=0.1) + 
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") + 
  geom_hline(yintercept = 250, linetype="dashed", colour = "red3") + 
  geom_hline(yintercept = 5000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of UMIs/Transcripts vs Number of Cells") +   
  stat_smooth(method=lm) + 
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() + 
  scale_colour_viridis_c() +
  facet_wrap(~sample)
#----- Visualize the correlation between genes detected and number of UMIs/transcripts 
# determine whether strong presence of cells with low numbers of genes/UMIs
seurat_combined@meta.data %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point(size=0.1) + 
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") + 
  geom_hline(yintercept = 250, linetype="dashed", colour = "red3") + 
  geom_hline(yintercept = 5000, linetype="dashed", colour = "red3") + 
  ggtitle("Number of UMIs/Transcripts vs Number of Cells") +   
  stat_smooth(method=lm) + 
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() + 
  scale_colour_viridis_c() +
  facet_wrap(~group)

#---------- Mitochondrial_counts_ratio
#----- Visualize the distribution of mitochondrial gene expression detected per cell
seurat_combined@meta.data %>% 
  ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 0.20, linetype="dashed", colour = "red3") + 
  ggtitle("Mitochondrial Counts Ratio") + 
  ylab("Cell density") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none")
#----- Visualize the distribution of mitochondrial gene expression detected per cell
seurat_combined@meta.data %>% 
  ggplot(aes(color=group, x=mitoRatio, fill=group)) + 
  geom_density(alpha = 0.2) + 
  geom_vline(xintercept = 0.20, linetype="dashed", colour = "red3") + 
  ggtitle("Mitochondrial Counts Ratio") + 
  ylab("Cell density") + 
  scale_x_log10() + 
  theme_classic()

#---------- Complexity
#----- Visualize the overall complexity of the gene expression by visualizing 
# the genes detected per UMI
seurat_combined@meta.data %>%
  ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) + 
  geom_density(alpha = 0.2) +
  geom_vline(xintercept = 0.85, linetype="dashed", colour = "red3") + 
  ggtitle("Complexity Genes per UMI") + 
  ylab("Cell density") + 
  theme_classic() + 
  theme(legend.position = "none")
#----- Visualize the overall complexity of the gene expression by visualizing 
# the genes detected per UMI
seurat_combined@meta.data %>%
  ggplot(aes(x=log10GenesPerUMI, color = group, fill=group)) + 
  geom_density(alpha = 0.2) +
  geom_vline(xintercept = 0.85, linetype="dashed", colour = "red3") + 
  ggtitle("Complexity Genes per UMI") + 
  ylab("Cell density") + 
  theme_classic() 
```

```{r Seurat_QC_Senescence}
#---------- Senescence Markers 
#----- Senescence marker vs Number of UMIs/Transcripts
seurat_combined@meta.data %>% 
  ggplot(aes(x=nUMI, y=CDKN2A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") +
  ggtitle("Cdkn2a Counts vs Number of UMIs/Transcripts") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)
seurat_combined@meta.data %>% 
  ggplot(aes(x=nUMI, y=CDKN1A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 500, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 20000, linetype="dashed", colour = "red3") +
  ggtitle("Cdkn1a Counts vs Number of UMIs/Transcripts") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)

#----- Senescence marker vs Number of Cells
seurat_combined@meta.data %>% 
  ggplot(aes(x=nGene, y=CDKN2A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 250, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 5000, linetype="dashed", colour = "red3") +
  ggtitle("Cdkn2a Counts vs Number of Cells") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)
seurat_combined@meta.data %>% 
  ggplot(aes(x=nGene, y=CDKN1A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 250, linetype="dashed", colour = "red3") + 
  geom_vline(xintercept = 5000, linetype="dashed", colour = "red3") +
  ggtitle("Cdkn1a Counts vs Number of Cells") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)

#----- Senescence marker vs mitoRatio
seurat_combined@meta.data %>% 
  ggplot(aes(x=mitoRatio, y=CDKN2A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 0.20, linetype="dashed", colour = "red3") + 
  ggtitle("Cdkn2a Counts vs Mitochondrial Counts Ratio") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)
seurat_combined@meta.data %>% 
  ggplot(aes(x=mitoRatio, y=CDKN1A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 0.20, linetype="dashed", colour = "red3") + 
  ggtitle("Cdkn1a Counts vs Mitochondrial Counts Ratio") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)

#----- Senescence marker vs Complexity Genes per UMI
seurat_combined@meta.data %>% 
  ggplot(aes(x=log10GenesPerUMI, y=CDKN2A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 0.85, linetype="dashed", colour = "red3") + 
  ggtitle("Cdkn2a Counts vs Complexity Genes per UMI") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)
seurat_combined@meta.data %>% 
  ggplot(aes(x=log10GenesPerUMI, y=CDKN1A_Count, color=sample)) + 
  geom_point(size=0.5) +
  geom_vline(xintercept = 0.85, linetype="dashed", colour = "red3") + 
  ggtitle("Cdkn1a Counts vs Complexity Genes per UMI") + 
  scale_x_log10() + 
  theme_classic() + 
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  theme(plot.title = element_text(hjust=0.5, face="bold")) + 
  facet_wrap(~sample)
```


## Seurat Filtering 
  
Remove poor quilty cells and doublets  
Keep  high quality cells without removing biologically relevant cells  
* nUMI >= 500  
* nUMI <= 20000  
* nGene >= 250  
* nGene <= 10000  
* mitoRatio < 0.2  
* log10GenesPerUMI > 0.85 

```{r Seurat_Filtering, eval=FALSE}
#---------- Seurat_Filtering
# Filter out low quality reads using selected thresholds - these will change with experiment
seurat_filtered <- subset(x = seurat_combined, 
                          subset = (nUMI >= 500) & 
                                   (nUMI <= 20000) & 
                                   (nGene >= 250) & 
                                   (nGene <= 10000) &
                                   (mitoRatio < 0.20) & 
                                   (log10GenesPerUMI > 0.85))
# seurat_filtered <- seurat_combined@meta.data %>% 
#                      filter(nUMI >= 500 & 
#                             nUMI <= 20000 & 
#                             nGene >= 250 & 
#                             nGene <= 10000 & 
#                             mitoRatio < 0.20 & 
#                             log10GenesPerUMI > 0.85)
# seurat_filtered <- subset(seurat_combined, cells = seurat_filtered$cells)
#---- save rds
saveRDS(seurat_filtered, file="seurat_results/seurat_filtered.rds")
```

```{r Seurat_Filtering_Figures}
#---------- Cell_Counts_Pre_and_Post_Filtered 
#----- Load data
seurat_filtered <- readRDS(file = "seurat_results/seurat_filtered.rds")
#----- Number of Cells
cbind(seurat_combined@meta.data$sample %>% 
        table() %>% 
        as.data.frame() %>% 
        dplyr::rename(Sample="."), 
      seurat_filtered@meta.data$sample %>% 
        table() %>% 
        as.data.frame() %>% 
        dplyr::rename(Sample=".")) %>% 
  kable(caption="Number of Cells", format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered")) %>% 
  add_header_above(c("Pre-Filtered" = 2, "Post-Filtered" = 2))

#----- Number of Cells
seurat_filtered@meta.data %>% 
  select(sample, group) %>% 
  dplyr::count(sample, group, sort = TRUE) %>% 
  kable(caption="Number of Filtered Cells", format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered"))

seurat_filtered@meta.data %>% 
  select(sample, group) %>% 
  dplyr::count(group, sort = TRUE) %>% 
  kable(caption="Number of Filtered Cells", format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered"))

#---------- Violin_plots
cowplot::plot_grid(VlnPlot(object = seurat_combined, 
                           features = c("nGene", "nUMI", "mitoRatio", "log10GenesPerUMI"), 
                           ncol = 1, 
                           pt.size=0, 
                           group.by="sample"), 
                   VlnPlot(object = seurat_filtered, 
                           features = c("nGene", "nUMI", "mitoRatio", "log10GenesPerUMI"), 
                           ncol = 1, 
                           pt.size=0, 
                           group.by="sample"), 
                   labels=c("Pre-Filtered", "Post-Filtered"), ncol = 2)

#---------- Scatter_plots
cowplot::plot_grid(FeatureScatter(object = seurat_combined, 
                                  feature1 = "nGene", 
                                  feature2 = "nUMI", 
                                  pt.size=0.01, 
                                  group.by="sample"), 
                   FeatureScatter(object = seurat_filtered, 
                                  feature1 = "nGene", 
                                  feature2 = "nUMI", 
                                  pt.size=0.01, 
                                  group.by="sample"),
                   FeatureScatter(object = seurat_combined, 
                                  feature1 = "nGene", 
                                  feature2 = "mitoRatio", 
                                  pt.size=0.01, 
                                  group.by="sample"), 
                   FeatureScatter(object = seurat_filtered, 
                                  feature1 = "nGene", 
                                  feature2 = "mitoRatio", 
                                  pt.size=0.01, 
                                  group.by="sample"), 
                   labels=c("Pre-Filtered", "Post-Filtered"), nrow=2, ncol=2)
#---- rm data
rm(seurat_combined)
```


## Seurat Cell Cycle Scoring 
  
Determine if cell cycle regression is required  
  
```{r Cell_Cycle_Scoring, eval=FALSE}
#---------- Cell_Cycle_Scoring
#----- Normalize the counts
seurat_phase <- NormalizeData(seurat_filtered)
#----- Load cell cycle markers
load(file="seurat_results/seurat_cycle.rda")

#----- Identify the most variable genescbin
seurat_phase <- FindVariableFeatures(seurat_phase, 
                     selection.method = "vst",
                     nfeatures = 2000, 
                     verbose = FALSE)
#----- Scale the counts
seurat_phase <- ScaleData(seurat_phase)
#----- Perform PCA
seurat_phase <- RunPCA(seurat_phase)
#----- Score cells for cell cycle
seurat_phase <- CellCycleScoring(seurat_phase, 
                                 g2m.features = g2m_genes, 
                                 s.features = s_genes)
#---- save rds
saveRDS(seurat_phase, file="seurat_results/seurat_phase.rds")
```

```{r Cell_Cycle_Scoring_PCA}
#---------- PCA
#----- Load data
seurat_phase <- readRDS(file = "seurat_results/seurat_phase.rds")
#----- Plot the PCA colored by cell cycle phase
DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")
# you can regress out the s and g2m score
# alternatively you could only regress out the s-g2m score to keep 
# non-cyling and cycling cell differences
DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "Phase",
        split.by = "group")
#----- Number of Cell Cycle Phase Cells
seurat_phase@meta.data %>% 
  dplyr::group_by(group) %>% 
  dplyr::count(Phase) %>%
  kable(caption="Number of Cell Cycle Phase Cells", format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered"))
#----- Remove data
rm(seurat_phase)
```


## Seurat Batch Correction
  
Standard batch correction is not considered standard anymore  
SCTransform batch correction allows for better normalization  
SCTransform method models the UMI counts using a regularized negative binomial  
model to remove the variation due to sequencing depth (total nUMIs per cell),  
while adjusting the variance based on pooling information across genes with   
similar abundances (similar to some bulk RNA-seq methods).  
  
If seurat_integrated has a memory error, run on the HPC  
conda env create --file r3.6_seurat.yml  
name: r3.6_seurat  
channels:  
  - conda-forge  
  - bioconda  
  - defaults  
dependencies:  
  - conda-forge::r-essentials=3.6  
  - conda-forge::libblas=3.8.0=14_mkl  
  - openblas  
  - r-doparallel  
  - conda-forge::r-seurat=3.2.3  
conda activate r3.6_seurat  
  
```{r Batch_Correction_SCTransform, eval=FALSE}
#---------- Batch_Correction_SCTransform
# The data is preprocessed by splitting into individual ‘assays’. 
# In this experiment, each individual sample is considered an ‘assay’.
# Ensure the dataset used here is the non-normalized dataset
DefaultAssay(seurat_filtered) <- "RNA"

#---------- Normalizing the data
# global-scaling normalization method “LogNormalize” that normalizes the feature expression 
# measurements for each cell by the total expression, multiplies this by a scale factor 
# (10,000 by default), and log-transforms the result. 
# Normalized values are stored in seurat_obj[["RNA"]]@data.
# identical(seurat_norm@assays$RNA@counts, seurat_norm@assays$RNA@data)
# Split seurat object by condition to perform cell cycle scoring and SCT
seurat_split <- SplitObject(seurat_filtered, split.by = "sample")
for (i in 1:length(seurat_split)) {
    seurat_split[[i]] <- NormalizeData(seurat_split[[i]], 
                                       verbose = TRUE)
    seurat_split[[i]] <- CellCycleScoring(seurat_split[[i]], 
                                          g2m.features=g2m_genes, 
                                          s.features=s_genes)
    seurat_split[[i]] <- SCTransform(seurat_split[[i]], 
                                     vars.to.regress = c("mitoRatio"))
}
#---------- Identification of highly variable features
# calculate a subset of features that exhibit high cell-to-cell variation in the dataset 
# (i.e, they are highly expressed in some cells, and lowly expressed in others)
# genes names stored in seurat_norm@assays[["RNA"]]@var.features
# variable features stored in seurat_norm@assays[["RNA"]]@meta.features
#----- Select the most variable features to use for integration
seurat_features <- SelectIntegrationFeatures(object.list = seurat_split, 
                                            nfeatures = 2000) 
#----- Prepare the SCT list object for integration
seurat_split <- PrepSCTIntegration(object.list = seurat_split, 
                                   anchor.features = seurat_features)
#----- modified to use young samples as the reference
# first time used reference=c(9,10,11,12,13,14,15,16)
seurat_anchors <- FindIntegrationAnchors(object.list = seurat_split, 
                                        normalization.method = "SCT", 
                                        anchor.features = seurat_features)
#---- Integrate across conditions
seurat_integrated <- IntegrateData(anchorset = seurat_anchors, 
                                   normalization.method = "SCT")
#---- Run PCA
seurat_integrated <- RunPCA(object = seurat_integrated)
ElbowPlot(object = seurat_integrated, 
          ndims = 50)
#---- Run UMAP
seurat_integrated <- RunUMAP(seurat_integrated, 
                             dims = 1:40,
			                       reduction = "pca")
#---- save rds
saveRDS(seurat_integrated, file="seurat_results/seurat_integrated.rds")
#---- Remove data
rm(seurat_filtered,seurat_split,seurat_features,seurat_anchors)
```

Harmony for batch integration  

```{r Batch_Correction_Harmony, eval=FALSE}
#---------- Batch_Correction_Harmony
#----- Load data
seurat_filtered <- readRDS(file = "seurat_results/seurat_filtered.rds")
# The data is preprocessed by splitting into individual ‘assays’. 
# In this experiment, each individual sample is considered an ‘assay’.
# Ensure the dataset used here is the non-normalized dataset
DefaultAssay(seurat_filtered) <- "RNA"

#---------- Normalizing the data
seurat_harmony <- NormalizeData(seurat_filtered) %>% 
                    FindVariableFeatures() %>% 
                    ScaleData() %>%  
                    RunPCA(verbose = FALSE)
# ElbowPlot(object = seurat_harmony, ndims = 50)

#---- Integrate across conditions
seurat_harmony <- RunHarmony(seurat_harmony, group.by.vars = "sample")
seurat_harmony <- RunUMAP(seurat_harmony, 
                          reduction = "harmony", 
                          dims = 1:30)
# DimPlot(seurat_harmony, split.by = "sample", ncol = 6) + NoLegend()

#---- save rds
saveRDS(seurat_harmony, file="seurat_results/seurat_harmony.rds")
#---- Remove data
rm(seurat_filtered)
```

```{r Batch_Correction_Harmony_Figures, warning=FALSE}
#----------  Plot PCA
#----- Load data
seurat_harmony <- readRDS(file = "seurat_results/seurat_harmony.rds")
#----- PCA
PCAPlot(seurat_harmony,
        group.by = "sample",
        split.by = "group") + NoLegend()
#----- Explore heatmap of PCs
DimHeatmap(seurat_harmony, 
           dims = 1:9, 
           cells = 500, 
           balanced = TRUE)
#----- Printing out the most variable genes driving PCs
print(x = seurat_harmony[["pca"]], 
      dims = 1:9, 
      nfeatures = 5)
#----- Plot the elbow plot
ElbowPlot(object = seurat_harmony, 
          ndims = 50)

#---------- Plot UMAP                             
DimPlot(seurat_harmony, 
        reduction = "umap", 
        split.by = "group") + NoLegend()
```


## Seurat Clustering_Cells

Seurat uses a graph-based clustering approach, which embeds cells in a graph  
structure, using a K-nearest neighbor (KNN) graph (by default), with edges  
drawn between cells with similar gene expression patterns.  

The resolution is an important argument that sets the “granularity” of the  
downstream clustering and will need to be optimized for every individual  
experiment. For datasets of 3,000 - 5,000 cells, the resolution set between   
0.4-1.4 generally yields good clustering. Increased resolution values lead to a  
greater number of clusters, which is often required for larger datasets. 

```{r Clustering_Cells, eval=FALSE}
#---------- Clustering_Cells
#----- Determine the K-nearest neighbor graph
#----- Determine the clusters for various resolutions                                 
seurat_cluster <- FindNeighbors(seurat_harmony, 
                                reduction = "harmony", 
                                dims = 1:30) %>% 
                  FindClusters(resolution = c(0.2, 0.4, 0.6, 0.8, 1.0, 1.2))
#---- save rds
saveRDS(seurat_cluster, file="seurat_results/seurat_cluster.rds")
#----- Remove data
rm(seurat_harmony)
```

```{r Clustering_Cells_UMAP, fig.cap="UMAP_1.2", out.width='100%'}
#---------- Explore resolutions
#----- Load data
load("seurat_cluster.RData")
#----- Assign identity of clusters
Idents(object = seurat_cluster) <- "RNA_snn_res.1.2"
#----- Plot UMAP                             
DimPlot(seurat_cluster, 
        reduction = "umap", 
        label = TRUE) + NoLegend()
```

Resolution 1.2  

```{r Clustering_Cells_nCells}
#----- Extract identity and sample information from seurat object to 
# determine the number of cells per cluster per sample
FetchData(seurat_cluster, vars = c("ident", "sample")) %>% 
  dplyr::count(ident, sample) %>% 
  tidyr::spread(ident, n) %>% 
  kable(caption="Number of Cells", 
        format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered"))
#----- UMAP of cells in each cluster by sample
DimPlot(seurat_cluster, 
        label = TRUE, 
        split.by = "group") + NoLegend()
#----- plot
VlnPlot(object = seurat_cluster, 
        features = c("nGene", "nUMI", "log10GenesPerUMI", "mitoRatio"), 
        ncol = 1, 
        pt.size=0, 
        group.by="RNA_snn_res.1.2") 
FeaturePlot(seurat_cluster, 
            reduction = "umap", 
            features = c("nUMI", "nGene", "log10GenesPerUMI", "mitoRatio"),
            ncol = 2, 
            pt.size = 0.4, 
            sort.cell = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
```

```{r Clustering_Cells_phase, eval=FALSE}
#----- cell phases
FetchData(seurat_cluster, vars = c("ident", "Phase")) %>% 
  dplyr::count(ident, Phase) %>% 
  tidyr::spread(ident, n) %>% 
  kable(caption="Number of Cells", format.args = list(big.mark = ",")) %>% 
  kable_styling(c("striped", "bordered"))
#----- Explore whether clusters segregate by cell cycle phase
DimPlot(seurat_cluster,
        label = TRUE, 
        split.by = "Phase")  + NoLegend()
#----- plot
FeaturePlot(seurat_cluster, 
            reduction = "umap", 
            features = c("S.Score", "G2M.Score"),
            ncol = 2,
            pt.size = 0.4, 
            sort.cell = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
```

```{r Clustering_Cells_PCs}
#----- Defining the information in the seurat object of interest
# Extracting this data from the seurat object
umap_data <- FetchData(seurat_cluster, 
                     vars = c(paste0("PC_", 1:16), 
                              "ident", "UMAP_1", "UMAP_2"))
#----- Extract the UMAP coordinates for the first 10 cells
#seurat_cluster@reductions$umap@cell.embeddings[1:10, 1:2]
# Adding cluster label to center of cluster on UMAP
umap_label <- FetchData(seurat_cluster, 
                        vars = c("ident", "UMAP_1", "UMAP_2"))  %>%
               group_by(ident) %>%
               summarise(x=mean(UMAP_1), y=mean(UMAP_2))
#----- Plotting a UMAP plot for each of the PCs
map(paste0("PC_", 1:16), function(pc){
        ggplot(umap_data, 
               aes(UMAP_1, UMAP_2)) +
                geom_point(aes_string(color=pc), 
                           alpha = 0.7) +
                scale_color_gradient(guide = FALSE, 
                                     low = "grey90", 
                                     high = "blue")  +
                geom_text(data=umap_label, 
                          aes(label=ident, x, y)) +
                ggtitle(pc)
}) %>% 
        plot_grid(plotlist = .)
#----- Explore heatmap of PCs
DimHeatmap(seurat_cluster, 
           dims = 1:9, 
           cells = 500, 
           balanced = TRUE)
#----- Printing out the most variable genes driving PCs
print(x = seurat_cluster[["pca"]], 
      dims = 1:9, 
      nfeatures = 5)
```


## Seurat Marker_Identification

```{r Marker_Identification, eval=FALSE}
#---------- Marker_Identification
#----- Select the RNA counts slot to be the default assay
DefaultAssay(seurat_cluster) <- "RNA"
#----- Normalize RNA data for visualization purposes
seurat_markers <- NormalizeData(seurat_cluster, verbose = FALSE)
#---- save rds
saveRDS(seurat_markers, file="seurat_results/seurat_markers.rds")
```

```{r Marker_Identification_Paper}
#---------- Marker_Identification
# https://panglaodb.se/markers.html?cell_type=%27Astrocytes%27
# http://bio-bigdata.hrbmu.edu.cn/CellMarker/search.jsp?species=Human&tissue=Brain&cellname=Astrocyte
#----- Load data
seurat_markers <- readRDS(file = "seurat_results/seurat_markers.rds")
#----- remove the x-axis text and tick
# plot.margin to adjust the white space between each plot.
# ... pass any arguments to VlnPlot in Seurat
vlnplot_gg <- function(obj, 
                       feature, 
                       pt.size = 0, 
                       plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                       ...){
  p <- VlnPlot(obj, features = feature, pt.size = pt.size, ... ) + 
    xlab("") + 
    ylab(feature) + 
    ggtitle("") + 
    theme(legend.position = "none", 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.title.y = element_text(size = rel(1), angle = 0), 
          axis.text.y = element_text(size = rel(1)), 
          plot.margin = plot.margin ) 
  return(p)
}
#----- extract the max value of the y axis
vlnplot_max<- function(p){
  ymax<- max(ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)
  return(ceiling(ymax))
}
#----- main function
vlnplot_stacked <- function(obj, features,
                          pt.size = 0, 
                          plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                          ...){
  plot_list<- purrr::map(features, function(x) vlnplot_gg(obj = obj,
                                                          feature = x, ...))
  # Add back x-axis title to bottom plot. patchwork is going to support this?
  plot_list[[length(plot_list)]] <- plot_list[[length(plot_list)]] + 
                                      theme(axis.text.x=element_text(), 
                                            axis.ticks.x = element_line())
  # change the y-axis tick to only max value 
  ymaxs <- purrr::map_dbl(plot_list, vlnplot_max)
  plot_list <- purrr::map2(plot_list, ymaxs, function(x,y) x + 
                             scale_y_continuous(breaks = c(y)) + 
                             expand_limits(y = y))
  p <- patchwork::wrap_plots(plotlist = plot_list, ncol = 1)
  return(p)
}

#---------- UMAP
DimPlot(seurat_markers, reduction="umap", label = TRUE) + NoLegend()
# rownames(seurat_markers@assays$RNA)[grepl(pattern = "CD3", rownames(seurat_markers@assays$RNA))]

#---------- Astrocyte_Cells
# GFAP, AQP4 (6,16)
vlnplot_stacked(obj = seurat_markers, 
               features = c("GFAP", "AQP4"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("GFAP", "AQP4"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Astrocyte_Restricted_Precursors_Cells
# CD44 ()
vlnplot_stacked(obj = seurat_markers, 
                features = c("CD44"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("CD44"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- B_Cells
# CD27,CD38,CD24,CD79A ()
vlnplot_stacked(obj = seurat_markers, 
                features = c("CD27", "CD38", "CD24", "CD79A"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("CD27", "CD38", "CD24", "CD79A"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Choroid_Plexus_Cells
# KL, CHMP1A, TTR, SLC26A11 ()
vlnplot_stacked(obj = seurat_markers, 
                features = c("KL", "CHMP1A", "SLC26A11"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("KL", "CHMP1A", "SLC26A11"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Dopaminergic_Neurons
# TH ()
# vlnplot_stacked(obj = seurat_markers, 
#                features = c("TH"))
# FeaturePlot(seurat_markers, 
#            reduction = "umap",
#            features = c("TH"), 
#            min.cutoff = "q10", 
#            max.cutoff = "q90",
#            label = TRUE)

#---------- Endothelial_Cells
# FLT1, DUSP1 (27)
vlnplot_stacked(obj = seurat_markers, 
               features = c("FLT1", "DUSP1"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("FLT1", "DUSP1"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Ependymal_Cells
# FOXJ1 ()
# vlnplot_stacked(obj = seurat_markers, 
#                features = c("FOXJ1"))
# FeaturePlot(seurat_markers, 
#            reduction = "umap",
#            features = c("FOXJ1"), 
#            min.cutoff = "q10", 
#            max.cutoff = "q90",
#            label = TRUE)

#---------- Excitatory_Neurons (Glutaminergic neurons)
# SATB2, SLC17A6 (2,3,4,7,8,12,14,15,18,19,20,21,23,25,28,30,31)
vlnplot_stacked(obj = seurat_markers, 
               features = c("SATB2", "SLC17A7", "SLC17A7"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("SATB2", "SLC17A6", "SLC17A7"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Inhibitory_Neurons (GABAergic neurons)
# GAD1, GAD2 (9,10,11,17,22,24,26)
vlnplot_stacked(obj = seurat_markers, 
               features = c("GAD1", "GAD2"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("GAD1", "GAD2"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Macrophage_Cells
# CD68 TREM2 ()
vlnplot_stacked(obj = seurat_markers, 
               features = c("CD68", "TREM2"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("CD68", "TREM2"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Microglial_Cells
# PLXDC2, CSF1R (13,29)
vlnplot_stacked(obj = seurat_markers, 
               features = c("APBB1IP", "CSF1R")) 
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("APBB1IP", "CSF1R"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Monocyte_Cells
# CD40 CXCR4 ()
vlnplot_stacked(obj = seurat_markers, 
               features = c("CD40", "CXCR4"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("CD40", "CXCR4"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Neural_Stem_Cells
# SOX2 SOX9 ()
vlnplot_stacked(obj = seurat_markers, 
                features = c("SOX2", "SOX9"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("SOX2", "SOX9"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Oligodendrocyte_Cells
# MOG, MOBP (0,1)
vlnplot_stacked(obj = seurat_markers, 
               features = c("MOG", "MOBP")) 
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("MOG", "MOBP"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Oligodendrocyte_Precursor_Cells
# OLIG1, VCAN (5)
vlnplot_stacked(obj = seurat_markers, 
               features = c("OLIG1", "VCAN"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("OLIG1", "VCAN"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- Pericyte_Cells
# PDGFRB ()
vlnplot_stacked(obj = seurat_markers, 
               features = c("PDGFRB"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("PDGFRB"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)

#---------- T_Cells
# CCR6 CD4 ()
vlnplot_stacked(obj = seurat_markers, 
               features = c("CCR6", "CD4"))
FeaturePlot(seurat_markers, 
            reduction = "umap",
            features = c("CCR6", "CD4"), 
            min.cutoff = "q10", 
            max.cutoff = "q90",
            label = TRUE)
```


## Seurat Cluster_Annotation

```{r Cluster_Annotation, eval=FALSE}
#---------- Cluster_Annotation
#----- Annotate each cluster
seurat_anno <- RenameIdents(seurat_markers, 
                            `0` = "Oligodendrocyte_Cells",
                            `1` = "Oligodendrocyte_Cells",
                            `2` = "Excitatory_Neurons",
                            `3` = "Excitatory_Neurons",
                            `4` = "Excitatory_Neurons",
                            `5` = "Oligodendrocyte_Precursor_Cells",
                            `6` = "Astrocyte_Cells",
                            `7` = "Excitatory_Neurons",
                            `8` = "Excitatory_Neurons",
                            `9` = "Inhibitory_Neurons",
                            `10` = "Inhibitory_Neurons",
                            `11` = "Inhibitory_Neurons",
                            `12` = "Excitatory_Neurons",
                            `13` = "Microglial_Cells",
                            `14` = "Excitatory_Neurons",
                            `15` = "Excitatory_Neurons",
                            `16` = "Astrocyte_Cells",
                            `17` = "Inhibitory_Neurons",
                            `18` = "Excitatory_Neurons",
                            `19` = "Excitatory_Neurons",
                            `20` = "Excitatory_Neurons",
                            `21` = "Excitatory_Neurons",
                            `22` = "Inhibitory_Neurons",
                            `23` = "Excitatory_Neurons",
                            `24` = "Inhibitory_Neurons",
                            `25` = "Excitatory_Neurons",
                            `26` = "Inhibitory_Neurons",
                            `27` = "Endothelial_Cells",
                            `28` = "Excitatory_Neurons",
                            `29` = "Microglial_Cells",
                            `30` = "Excitatory_Neurons",
                            `31` = "Excitatory_Neurons")
#----- save these annotations in meta.data
seurat_anno@meta.data$cluster_annotation <- Idents(seurat_anno)

#---------- Manually Edit Annotations
#----- UMAP
DimPlot(seurat_anno, 
        reduction="umap", 
        label = TRUE) + NoLegend()
#----- Metadata
metadata <- seurat_anno@meta.data
metadata$cluster_annotation <- "Na"
#----- Edit
umap <- DimPlot(seurat_anno, reduction="umap", label = TRUE) 
anno <- CellSelector(plot = umap)
metadata$cluster_annotation[metadata$cells %in% anno] <- "Oligodendrocyte_Precursor_Cells"
#----- Metadata
seurat_anno@meta.data <- metadata
#----- Assign identity of clusters
Idents(object = seurat_anno) <- "cluster_annotation"
#----- UMAP
DimPlot(seurat_anno, 
        reduction="umap", 
        label = TRUE) + NoLegend()
#---- save rds
saveRDS(seurat_anno, file="seurat_results/seurat_anno.rds")
#----- Remove data
rm(seurat_markers,metadata,umap,anno)
```

```{r Cluster_Annotation_plots}
#---------- Cluster_Annotation
#----- Load data
seurat_anno <- readRDS(file = "seurat_results/seurat_anno.rds")
#-----  UMAP cell clusters
DimPlot(seurat_anno, 
        reduction="umap", 
        label = FALSE)
DimPlot(seurat_anno, 
        reduction="umap", 
        label = TRUE, 
        label.size = 3, 
        repel = TRUE) + NoLegend()
DimPlot(seurat_anno, 
        reduction="umap", 
        label = TRUE,
        split.by = "group",
        label.size = 3, 
        repel = TRUE) + NoLegend()
#----- Celltype Counts
table(seurat_anno@meta.data$cluster_annotation)
#----- may want to ask if we are truely seperating interneurons from excitatory
DotPlot(seurat_anno,
        features = c("GFAP",    # Astrocyte_Cells
                     "DUSP1",   # Endothelial_Cells
                     "SLC17A7", # Excitatory_Neurons
                     "GAD1",    # Inhibitory_Neurons
                     "APBB1IP", # Microglial_Cells
                     "MOBP",    # Oligodendrocyte_Cells
                     "VCAN"),  # Oligodendrocyte_Precursor_Cells
        cols = c("blue", "red"), 
        dot.scale = 8) + 
  RotatedAxis()
# If we want to remove a cell cluster
#seurat_subset <- subset(seurat_integrated, idents = "Endothelial_cells", invert = TRUE)

#---------- Number_of_Cells
#----- Sample
seurat_anno@meta.data %>% 
  dplyr::count(cluster_annotation, sample) %>%  
  ggplot(aes(x=cluster_annotation, y=n, fill=sample)) + 
    geom_bar(position="dodge", stat="identity") + 
    ggtitle("Number of Cells") + 
    ylab("Cell Count") +
    theme_classic() + 
    theme(legend.position = "none") + 
    facet_wrap(~cluster_annotation, scale="free", ncol = 4)
#----- Group
seurat_anno@meta.data %>% 
  dplyr::count(cluster_annotation, group) %>%  
  ggplot(aes(x=cluster_annotation, y=n, fill=group)) + 
    geom_bar(position="dodge", stat="identity") + 
    ggtitle("Number of Cells") + 
    ylab("Cell Count") +
    theme_classic() + 
    scale_fill_brewer(palette="Dark2") +
    theme(legend.position = "none") + 
    facet_wrap(~cluster_annotation, scale="free", ncol = 4)

#---------- Frequency_of_Cells
#----- Sample
seurat_anno@meta.data %>% 
  dplyr::count(cluster_annotation, sample) %>% 
  group_by(sample) %>% 
  mutate(freq = n / sum(n)) %>% 
  ggplot(aes(x=cluster_annotation, y=freq, fill=sample)) + 
    geom_bar(position="dodge", stat="identity") + 
  	ggtitle("Frequency of Cells") + 
    theme_classic() + 
    theme(legend.position = "none") + 
  	# theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
    facet_wrap(~cluster_annotation, scale="free", ncol = 4)
#----- Group
seurat_anno@meta.data %>% 
  dplyr::count(cluster_annotation, group) %>% 
  group_by(group) %>% 
  mutate(freq = n / sum(n)) %>% 
  ggplot(aes(x=cluster_annotation, y=freq, fill=group)) + 
    geom_bar(position="dodge", stat="identity") + 
  	ggtitle("Frequency of Cells") + 
    theme_classic() + 
    scale_fill_brewer(palette="Dark2") +
    theme(legend.position = "none") + 
  	# theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
    facet_wrap(~cluster_annotation, scale="free", ncol = 4)


```






